pkg/controller/staticpod/upgradeinfo/upgrade_info.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2023 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgradeinfo
+
+import (
+	"context"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/kubectl/pkg/util/podutils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "github.com/openyurtio/openyurt/pkg/apis/apps/v1alpha1"
+	"github.com/openyurtio/openyurt/pkg/controller/staticpod/util"
+)
+
+const (
+	StaticPodHashAnnotation     = "openyurt.io/static-pod-hash"
+	OTALatestManifestAnnotation = "openyurt.io/ota-latest-version"
+)
+
+// UpgradeInfo is a structure that stores some information used by static pods to upgrade.
+type UpgradeInfo struct {
+	// Static pod running on the node
+	StaticPod *corev1.Pod
+
+	// Upgrade worker pod running on the node
+	WorkerPod *corev1.Pod
+
+	// Indicate whether the static pod on the node needs to be upgraded.
+	// If true, the static pod is not up-to-date and needs to be upgraded.
+	UpgradeNeeded bool
+
+	// Indicate whether the worker pod on the node is running.
+	// If true, then the upgrade operation is in progress and does not
+	// need to create a new worker pod.
+	WorkerPodRunning bool
+
+	// Indicate whether the node is ready. It's used in Auto mode.
+	Ready bool
+}
+
+// New constructs the upgrade information for nodes which have the target static pod
+func New(c client.Client, instance *appsv1alpha1.StaticPod, workerPodName string) (map[string]*UpgradeInfo, error) {
+	infos := make(map[string]*UpgradeInfo)
+
+	// upgrade worker pod is default in "kube-system" namespace which may be different with target static pod's namespace
+	var podList, systemPodList corev1.PodList
+	if err := c.List(context.TODO(), &podList, &client.ListOptions{Namespace: instance.Namespace}); err != nil {
+		return nil, err
+	}
+
+	if err := c.List(context.TODO(), &systemPodList, &client.ListOptions{Namespace: instance.Namespace}); err != nil {
```

Comment:
I think we need to list upgrade worker pods in `kube-system` namespace instead of `instance.Namespace`